[SPARK-14743][YARN] Add a configurable credential manager for Spark running on YARN

jerryshao opened the pull request:
## What changes were proposed in this pull request?

Add a configurable token manager for Spark on running on yarn.
### Current Problems
1. Supported token provider is hard-coded, currently only hdfs, hbase and hive are supported and it is impossible for user to add new token provider without code changes.
2. Also this problem exits in timely token renewer and updater.
### Changes In This Proposal

In this proposal, to address the problems mentioned above and make the current code more cleaner and easier to understand, mainly has 3 changes:
1. Abstract a `ServiceTokenProvider` as well as `ServiceTokenRenewable` interface for token provider. Each service wants to communicate with Spark through token way needs to implement this interface. 
2. Provide a `ConfigurableTokenManager` to manage all the register token providers, also token renewer and updater. Also this class offers the API for other modules to obtain tokens, get renewal interval and so on.
3. Implement 3 built-in token providers `HDFSTokenProvider`, `HiveTokenProvider` and `HBaseTokenProvider` to keep the same semantics as supported today. Whether to load in these built-in token providers is controlled by configuration "spark.yarn.security.tokens.${service}.enabled", by default for all the built-in token providers are loaded.
### Behavior Changes

For the end user there's no behavior change, we still use the same configuration `spark.yarn.security.tokens.${service}.enabled` to decide which token provider is enabled (hbase or hive).

For user implemented token provider (assume the name of token provider is "test") needs to add into this class should have two configurations:
1. `spark.yarn.security.tokens.test.enabled` to true
2. `spark.yarn.security.tokens.test.class` to the full qualified class name.

So we still keep the same semantics as current code while add one new configuration.
### Current Status
- [x] token provider interface and management framework.
- [x] implement built-in token providers (hdfs, hbase, hive).
- [x] Coverage of unit test.
- [x] Integrated test with security cluster.
## How was this patch tested?

Unit test and integrated test.

Please suggest and review, any comment is greatly appreciated.

jerryshao committed:
Add ConfigurableTokenManager initial commit

jerryshao committed:
Further change on ConfigurableTokenManager

jerryshao committed:
Some refactory works and unit test added

jerryshao committed:
Add more unit tests

jerryshao committed:
Add more comments

SparkQA commented:
**[Test build #61830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61830/consoleFull)** for PR 14065 at commit [`90f194e`](https://github.com/apache/spark/commit/90f194e34ffd198d2b1ee5b04f24afd8c4454d90).

SparkQA commented:
**[Test build #61830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61830/consoleFull)** for PR 14065 at commit [`90f194e`](https://github.com/apache/spark/commit/90f194e34ffd198d2b1ee5b04f24afd8c4454d90).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61830/
Test FAILed.

jerryshao committed:
Add Mima exclusion

SparkQA commented:
**[Test build #61896 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61896/consoleFull)** for PR 14065 at commit [`2b17f2e`](https://github.com/apache/spark/commit/2b17f2e906b2fd67e1c2398fd8ba88b64b797550).

SparkQA commented:
**[Test build #61896 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61896/consoleFull)** for PR 14065 at commit [`2b17f2e`](https://github.com/apache/spark/commit/2b17f2e906b2fd67e1c2398fd8ba88b64b797550).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61896/
Test PASSed.

jerryshao committed:
supress the token fetch exception

SparkQA commented:
**[Test build #61912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61912/consoleFull)** for PR 14065 at commit [`c9d9ed0`](https://github.com/apache/spark/commit/c9d9ed0cd0aef6b8017c04e635f27ef123a48887).

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/AMDelegationTokenRenewer.scala:
```diff
@@ -171,10 +174,9 @@ private[yarn] class AMDelegationTokenRenewer(
-        hadoopUtil.obtainTokenForHiveMetastore(sparkConf, freshHadoopConf, tempCreds)
-        hadoopUtil.obtainTokenForHBase(sparkConf, freshHadoopConf, tempCreds)
+        hdfsTokenProvider(sparkConf).setNameNodesToAccess(sparkConf, Set(dst))
+        hdfsTokenProvider(sparkConf).setTokenRenewer(null)
```
why are we setting this to null here?  is it supposed to indicate that its not supposed to be renewed internally?

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/HDFSTokenProvider.scala:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  override def getTokenRenewalInterval(sparkConf: SparkConf, serviceConf: Configuration): Long = {
+    // We cannot use the tokens generated above since those have renewer yarn. Trying to renew
```
update comment to be more specific then "above" since this has moved

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -390,8 +390,9 @@ private[spark] class Client(
     val fs = destDir.getFileSystem(hadoopConf)
-    val nns = YarnSparkHadoopUtil.get.getNameNodesToAccess(sparkConf) + destDir
-    YarnSparkHadoopUtil.get.obtainTokensForNamenodes(nns, hadoopConf, credentials)
+    hdfsTokenProvider(sparkConf).setNameNodesToAccess(sparkConf, Set(destDir))
```
it seems a bit odd to me that we are doing these extra things for hdfs outside of the token provider.

what happens if a user needs to do something similar for some other service they want to implement their own class to handle?

vanzin replied:
+1; it would be better if all interactions with token providers were done through the common interface; it seems like these HDFS-specific calls could easily be moved to the HDFS token provider.

jerryshao replied:
Yes, I agreed that it is a bit odd for HDFS token provider to have special code path, let me find out a way to get rid of this.

tgravescs commented:
I took a quick look through.

It might be nice to think about how we could handle other credentials.

For instance Apache Kafka currently doesn't have tokens so you need keytab or TGT and jaas conf file. Yes they are adding tokens but in in the mean time how does that work.  Are there other services similar to that.

Can we handle things other then Tokens?  it does appear that I could implement my own ServiceTokenProvider that goes off to really any service and I can put things into the Credentials object as Token or as a Secret so perhaps we are covered here.  But perhaps that means we should rename things to be obtainCredentials rather then obtainTokens.

Are there specific services you were thinking about here? We could atleast use those as examples to make sure interface fits those.

SparkQA commented:
**[Test build #61912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61912/consoleFull)** for PR 14065 at commit [`c9d9ed0`](https://github.com/apache/spark/commit/c9d9ed0cd0aef6b8017c04e635f27ef123a48887).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61912/
Test FAILed.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala:
```diff
@@ -96,237 +87,19 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
   private[spark] override def startExecutorDelegationTokenRenewer(sparkConf: SparkConf): Unit = {
-    tokenRenewer = Some(new ExecutorDelegationTokenUpdater(sparkConf, conf))
-    tokenRenewer.get.updateCredentialsIfRequired()
+    configurableTokenManager(sparkConf).delegationTokenUpdater(conf)
```
I find this syntax a little confusing. You're calling `configurableTokenManager(sparkConf)` in a bunch of different places. To me that looks like either:

- each call is creating a new token manager
- there's some cache of token managers somewhere keyed by the spark configuration passed here

Neither sounds good to me. And the actual implementation is actually neither: there's a single token manager singleton that is instantiated in the first call to `configurableTokenManager`.

Why doesn't `Client` instantiate a token manager in its constructor instead? Another option is to have an explicit method in `ConfigurableTokenManager` to initialize the singleton, although I'm not a fan of singletons in general.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ConfigurableTokenManager.scala:
```diff
@@ -0,0 +1,214 @@
+  // Maintain all the registered token providers
+  private val tokenProviders = mutable.HashMap[String, ServiceTokenProvider]()
+
+  private val defaultTokenProviders = Map(
```
I'd rather use `java.util.ServiceLoader` for this. You'll need something like that at some point anyway, to support other token providers. Doing that now has the extra benefit of using the same code for built-in and third party providers.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ConfigurableTokenManager.scala:
```diff
@@ -0,0 +1,214 @@
+  // executor side.
+  private var _delegationTokenUpdater: ExecutorDelegationTokenUpdater = null
+
+  def initialize(): Unit = {
```
A lot of this method would go away by using `java.util.ServiceLoader`.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ConfigurableTokenManager.scala:
```diff
@@ -0,0 +1,214 @@
+   */
+  def obtainTokens(conf: Configuration, creds: Credentials): Array[Token[_]] = {
+    val tokenBuf = mutable.ArrayBuffer[Token[_]]()
+    tokenProviders.values.foreach { provider =>
```
minor: you can use `flatMap` instead of a mutable buffer.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ConfigurableTokenManager.scala:
```diff
@@ -0,0 +1,214 @@
+  def obtainTokens(conf: Configuration, creds: Credentials): Array[Token[_]] = {
+    val tokenBuf = mutable.ArrayBuffer[Token[_]]()
+    tokenProviders.values.foreach { provider =>
+      if (provider.isTokenRequired(conf)) {
```
Seems like you could remote providers that don't require tokens during `initialize` if you make a `SparkConf` available to it. Then you could avoid passing `SparkConf` to all of these methods.

You could also make `Configuration` a constructor args for the same reasons.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ServiceTokenProvider.scala:
```diff
@@ -0,0 +1,75 @@
+   */
+  def obtainTokensFromService(
+      sparkConf: SparkConf,
+      serviceConf: Configuration,
```
Note the name here is misleading. It won't be the service's conf, but really a `YarnConfiguration`. Note how both Hive and HBase providers have to load their own configuration to be able to see service-specific settings.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ServiceTokenProvider.scala:
```diff
@@ -0,0 +1,75 @@
+  /**
+   * Get the time length from now to next renewal.
+   */
+  def getTimeFromNowToRenewal(
```
You only really need this method in the interface, right? The token provider should know what info it needs to calculate this value. It might not event need `getTokenRenewalInterval` for that (HDFS does, but then that logic should live inside the HDFS provider).

At that point, you could just merge both interfaces and have this method return an `Option` (None = no renewal necessary), or some magic value (e.g. `-1`) to indicate no renewal.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/token/ServiceTokenProvider.scala:
```diff
@@ -0,0 +1,75 @@
+  /**
+   *  Obtain tokens from this service, tokens will be added into Credentials and return as array.
+   */
+  def obtainTokensFromService(
```
If you follow Tom's suggestion and turn this into a generic "obtainCredentials" method, then you could potentially merge it with `getTimeFromNowToRenewal` too.

e.g. the provider is responsible for adding the tokens to the `Credentials` object, and it returns when it should be called again to renew those tokens (or obtain new credentials). One less method in the interface!

jerryshao replied:
This looks much more consolidate, I will change to this, thanks.

jerryshao commented:
Thanks a lot @tgravescs and @vanzin for your suggestions, I will change the codes accordingly, greatly appreciate your comments.

jerryshao committed:
Address the comments to change consolidate the interface and change the logics

jerryshao committed:
Add default credential providers for service loader

jerryshao committed:
Add some comments

SparkQA commented:
**[Test build #62148 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62148/consoleFull)** for PR 14065 at commit [`0fbf25b`](https://github.com/apache/spark/commit/0fbf25b51285ea433a3dcf733b92a1d785b3d017).

SparkQA commented:
**[Test build #62148 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62148/consoleFull)** for PR 14065 at commit [`0fbf25b`](https://github.com/apache/spark/commit/0fbf25b51285ea433a3dcf733b92a1d785b3d017).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62148/
Test FAILed.

jerryshao committed:
Add rat exclude

SparkQA commented:
**[Test build #62149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62149/consoleFull)** for PR 14065 at commit [`60a275f`](https://github.com/apache/spark/commit/60a275f9487eacab8ce8c70d6917d3aebe16d131).

jerryshao commented:
@tgravescs and @vanzin , these days I did some code refactoring work on this patch. Here listed changes I did compared to previous code:

1. Change the interface `ServiceTokenProvider` to `ServiceCredentialProvider` with the main method changed to `obtainCredentials` as suggested in comments.

    Now since we're not limiting to tokens, so instead of obtaining tokens, here changed to obtain credentials. And the method `obtainCredentials` is defined as:
    
    ```scala
     def obtainCredentials(hadoopConf: Configuration, creds: Credentials): Option[Long]
    ```
   
    Here the return value `Option[Long]` means the time of next renewal, return `Some(Long)` if this credential is renewable, otherwise returns `None`.
  
    Also remove several redundant methods like get token renewal interval and so on.

2. Change `ConfigurableTokenManager` to `ConfigurableCredentialManager` to manager all the credential providers.
3. Change the way to load credential providers to ServiceLoader as suggested in comments.
4. Change initialization way from singleton to normal way.
6. Change the mechanism of checking credentials in `AMDelegationRenewer` and `ExecutorDelegationTokenUpdate`. Since now we can get the time of next renewal, so we use this to decide when to wake up to check the new credentials.

Please help to review, thanks a lot for your time and greatly appreciate your comments.

jerryshao commented:
Also thinking about one example to land this feature, I think Kafka might be one candidate, they also have delegation token based proposal [KIP-48](https://cwiki.apache.org/confluence/display/KAFKA/KIP-48+Delegation+token+support+for+Kafka).

SparkQA commented:
**[Test build #62149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62149/consoleFull)** for PR 14065 at commit [`60a275f`](https://github.com/apache/spark/commit/60a275f9487eacab8ce8c70d6917d3aebe16d131).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62149/
Test PASSed.

jerryshao committed:
Minor change

SparkQA commented:
**[Test build #62219 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62219/consoleFull)** for PR 14065 at commit [`a8141a5`](https://github.com/apache/spark/commit/a8141a54dd87f77c3ce6eb0bccfabacd64841b4b).

SparkQA commented:
**[Test build #62219 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62219/consoleFull)** for PR 14065 at commit [`a8141a5`](https://github.com/apache/spark/commit/a8141a54dd87f77c3ce6eb0bccfabacd64841b4b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62219/
Test FAILed.

vanzin commented on dev/.rat-excludes:
```diff
@@ -99,3 +99,4 @@ spark-deps-.*
 .*tsv
 org.apache.spark.scheduler.ExternalClusterManager
 .*\.sql
+org.apache.spark.deploy.yarn.security.ServiceCredentialProvider
```
Instead of adding these files like this, isn't it possible to just say `META-INF/services/*`?

jerryshao replied:
I followed the convention of other services like `DataSource` which also needs to be excluded from rat check.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -390,8 +393,22 @@ private[spark] class Client(
+
+    // Merge credentials obtained from registered providers
+    var nearestTimeOfNextRenewal = Long.MaxValue
+    credentialManager.obtainCredentials(hadoopConf, credentials).foreach { t =>
```
You don't really need to calculate `nearestTimeOfNextRenewal` if `loginFromKeytab` is false, right?

Also, this all could be:

```
val nearestTimeOfNextRenewal =  credentialManager.obtainCredentials(hadoopConf, credentials).flatten.min
```

Just be careful to:
- validate that all returned times are after the current time
- watch out for when the result of `.flatten` is empty

jerryshao replied:
Yes, in the current implementation of `HDFSCredentialProvider`, time of next token renewal will only be calculated when principal is provided (`loginFromKeytab` is true).

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -69,6 +71,9 @@ private[yarn] class AMDelegationTokenRenewer(
     hadoopUtil.getConfBypassingFSCache(hadoopConf, new Path(credentialsFile).toUri.getScheme)
 
+  @volatile private var timeOfNextRenewal =
+    sparkConf.get(CREDENTIALS_RENEWAL_TIME).getOrElse(Long.MaxValue)
```
From my understanding you should never get to this code when  `CREDENTIALS_RENEWAL_TIME` is not set, so you should use `.get` to force and exception in case the config is not set.

If indeed that situation can occur and you really need a default value, it should be declared in the constant itself and not here where it's used.

jerryshao replied:
One possible scenario from current code is that if principal and keytab is configured, but all credential are non-renewable or disabled. In such case `CREDENTIALS_RENEWAL_TIME` may be empty but `AMDelegationTokenRenewer` will still be started because it only relies on principal configured or not.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -85,15 +90,14 @@ private[yarn] class AMDelegationTokenRenewer(
-      val renewalInterval = hadoopUtil.getTimeFromNowToRenewal(sparkConf, 0.75, credentials)
       // Run now!
-      if (renewalInterval <= 0) {
+      val leftTime = timeOfNextRenewal - System.currentTimeMillis()
```
s/leftTime/remainingTime

The previous name (renewalInterval) was also fine.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -171,10 +175,13 @@ private[yarn] class AMDelegationTokenRenewer(
-        hadoopUtil.obtainTokenForHiveMetastore(sparkConf, freshHadoopConf, tempCreds)
-        hadoopUtil.obtainTokenForHBase(sparkConf, freshHadoopConf, tempCreds)
+        var nearestNextTime = Long.MaxValue
+        credentialManager.obtainCredentials(freshHadoopConf, tempCreds).foreach { nextTime =>
```
This logic looks identical to the one in `Client.scala`. Maybe `credentialManager.obtainCredentials` should do this calculation and return when to next invoke credential renewal?

vanzin replied:
I also noticed that the logic will renew all tokens when the first token needs renewal. It's probably fine since this should happen very rarely (once every few days). But that also means that `obtainCredentialsFromService` is completely unused and can be deleted.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -193,8 +200,14 @@ private[yarn] class AMDelegationTokenRenewer(
       }
     }
     val nextSuffix = lastCredentialsFileSuffix + 1
+    // Time of next update should be late than renewal,
```
later?

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -193,8 +200,14 @@ private[yarn] class AMDelegationTokenRenewer(
     val nextSuffix = lastCredentialsFileSuffix + 1
+    // Time of next update should be late than renewal,
+    // which is around new token issue date + 0.8 * renew interval
+    val timeOfNextUpdate =
```
This is also part of the logic that could be consolidated in `credentialManager.obtainCredentials`. Also, comments above don't really match what the code is doing.

jerryshao replied:
Not sure how this part could be consolidate into `obtainCredentials` method, the return value of `obtainCredentials` is time of next renewal, here it is time of next updating, which should be slightly later than next renewal. 

So in shortly IIUC if we also want to consolidate this code, so we should change the return value `obtainCredentials` to `(timeOfNextRenewal, timeOfNextUpdate)`, what do you think?

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,158 @@
+  private val credentialProviders = mutable.HashMap[String, ServiceCredentialProvider]()
+
+  // Default crendetial providers that will be loaded automatically, unless specifically disabled.
+  private val defaultCredentialProviders = Map(
```
Why do you still need this now that the code is using `ServiceLoader`?

If `ServiceLoader.load` is not loading these providers, something is wrong.

jerryshao replied:
Sorry my fault, it is not necessary now, I will remove it.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,158 @@
+            c
+          }
+        }
+        .getOrElse(defaultCredentialProviders.keySet.find(_ == p.serviceName).isDefined.toString)
```
What's the purpose of this line? Seems like you want to enable internal providers and disable 3rd-party ones by default? Why?

jerryshao replied:
This means that by default hdfs, hive and hbase credential providers will be loaded unless explicitly disabled by configuration.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,158 @@
+        .getOrElse(defaultCredentialProviders.keySet.find(_ == p.serviceName).isDefined.toString)
+        .toBoolean
+    }.foreach {
+      p => credentialProviders(p.serviceName) = p
```
Don't use a mutable hash map. Instead:

```
.map {  p => (p.serviceName, p) }.toMap
```

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,158 @@
+    timeOfNextRenewals.toArray
+  }
+
+  def delegationTokenRenewer(hadoopConf: Configuration): AMDelegationTokenRenewer = synchronized {
```
This method has that weird singleton-like pattern again. Maybe it should just return a new renewer on every call, since it doesn't seem this class actually has any use for it.

vanzin replied:
You have two options here:

- pass the hadoop configuration in the manager's constructor, and lazily instantiate the token renewer (and updated) - a `lazy val` would suffice for that
- just return a new one for the given hadoop configuration and let the caller handle its lifecycle

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,158 @@
+    }
+  }
+
+  def delegationTokenUpdater(hadoopConf: Configuration
```
Same comment about the singleton-like interface.

jerryshao committed:
Address the comment

jerryshao committed:
Change to structure

SparkQA commented:
**[Test build #62467 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62467/consoleFull)** for PR 14065 at commit [`a66aed0`](https://github.com/apache/spark/commit/a66aed0d1f299b87f68f34f96f0ccc539507dfde).

jerryshao committed:
Further refactor

SparkQA commented:
**[Test build #62468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62468/consoleFull)** for PR 14065 at commit [`9863ca4`](https://github.com/apache/spark/commit/9863ca42427479bf79731f204f606d1ccae20491).

SparkQA commented:
**[Test build #62467 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62467/consoleFull)** for PR 14065 at commit [`a66aed0`](https://github.com/apache/spark/commit/a66aed0d1f299b87f68f34f96f0ccc539507dfde).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62467/
Test PASSed.

SparkQA commented:
**[Test build #62468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62468/consoleFull)** for PR 14065 at commit [`9863ca4`](https://github.com/apache/spark/commit/9863ca42427479bf79731f204f606d1ccae20491).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62468/
Test PASSed.

tgravescs commented:
sorry I haven't had time to review fully, taking a quick look I think it would be nice to add at least a blurb to the docs (running on yarn) about how to use this and pointer to scaladocs for ServiceCredentialProvider and the spark.yarn.security.credentials.%s.enabled configs and how things get loaded.

We also need to update docs for spark.yarn.security.credentials.%s.enabled.

jerryshao commented:
Thanks @tgravescs for your comments, I will add the docs about it.

jerryshao committed:
Further change the code and add docs

SparkQA commented:
**[Test build #62506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62506/consoleFull)** for PR 14065 at commit [`b8eeb28`](https://github.com/apache/spark/commit/b8eeb28b141b678cf4ccace36564f24536758132).

SparkQA commented:
**[Test build #62506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62506/consoleFull)** for PR 14065 at commit [`b8eeb28`](https://github.com/apache/spark/commit/b8eeb28b141b678cf4ccace36564f24536758132).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62506/
Test PASSed.

tgravescs commented on docs/running-on-yarn.md:
```diff
@@ -461,15 +461,28 @@ To use a custom metrics.properties for the application master and executors, upd
-  Currently supported services are: <code>hive</code>, <code>hbase</code>
+  Currently supported services are: <code>hdfs</code>, <code>hive</code>, <code>hbase</code>
+  <p/>
+  For other 3rd party services need to be communicate with Spark through security way should implement
```
a few grammar things here and I think we should separate this out into the "Running in a Secure Cluster" section.

It also has a few places that need updated from tokens. to credentials.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala:
```diff
@@ -319,6 +319,16 @@ package object config {
+    .timeConf(TimeUnit.MILLISECONDS)
+    .createWithDefault(Long.MaxValue)
+
+  private[spark] val CREDENTIALS_UPDATE_TIME = ConfigBuilder("spark.yarn.credentials.update")
```
so this is really executors fetching credentials from AM correct?  It took me a few minutes to figure out what this config was vs the one above, can we rename it  to spark.yarn.credentials.executors.updateTime and CREDENTIALS_EXECUTOR_UPDATE_TIME.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -390,8 +391,21 @@ private[spark] class Client(
+      nearestTimeOfNextRenewal != Long.MaxValue) {
+      sparkConf.set(CREDENTIALS_RENEWAL_TIME, nearestTimeOfNextRenewal)
+      val nearestTimeOfNextUpdate =
+        (nearestTimeOfNextRenewal - System.currentTimeMillis()) * 1.1 + System.currentTimeMillis()
```
what is the 1.1 here? just some time slightly in the future for executors to fetch?  Note that if they really return the time to renew from obtainTokens this could be to late.  With hdfs with return 75% of renew time. We either need to make sure to document the user should do something similar or we need to handle here.  Its probably safer to handle here.  The only downside would be if they make renewal interval really really small.  We should probably put warning/documentation in around obtain tokens about that too.

jerryshao replied:
Yes, this is a valid concern, I'm also thinking about this, I will try to improve this.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -400,8 +414,6 @@ private[spark] class Client(
     val distributedNames = new HashSet[String]
-    YarnSparkHadoopUtil.get.obtainTokenForHiveMetastore(sparkConf, hadoopConf, credentials)
-    YarnSparkHadoopUtil.get.obtainTokenForHBase(sparkConf, hadoopConf, credentials)
     if (credentials != null) {
```
this is now a bit out of place, the idea was to dump the tokens for debugging after they were filled.  We should move this after the obtainCredentials call above.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -193,10 +196,17 @@ private[yarn] class AMDelegationTokenRenewer(
     val nextSuffix = lastCredentialsFileSuffix + 1
+    // Time of next update should be later than renewal,
+    val timeOfNextUpdate =
+      (timeOfNextRenewal - System.currentTimeMillis()) * 1.1 + System.currentTimeMillis()
```
I'm a little concerned with the 1.1 here.  If renewal is every 24 hours, 1.1 times that is another 144 minutes past renewal, this is fine for hdfs which return 75% of renewal but its still quite long, it was 70 minutes before with 75% and 80% but I'm a little concerned with possibly other services and what the user could return.  So following my comment above perhaps we should move the logic from hdfs credentials into here and do like the 75%/80% (or 85%) again.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -171,10 +174,10 @@ private[yarn] class AMDelegationTokenRenewer(
+        val nearestNextTime = credentialManager.obtainCredentials(freshHadoopConf, tempCreds)
+        require(nearestNextTime > System.currentTimeMillis(),
+          s"Time of next renewal $nearestNextTime is earlier than now")
+        timeOfNextRenewal = nearestNextTime
```
driverTokenRenewerRunnable is hardcoded to renew every hour: 
              delegationTokenRenewer.schedule(this, 1, TimeUnit.HOURS)

We should probably make this configurable.  Ideally nothing needs to be renewed that often but if we are wanting to support arbitrary services making it configurable would make sense.

tgravescs replied:
Nevermind, ignore this I misread the code... that is only if exception occurs

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMDelegationTokenRenewer.scala:
```diff
@@ -171,10 +174,10 @@ private[yarn] class AMDelegationTokenRenewer(
-        hadoopUtil.obtainTokenForHiveMetastore(sparkConf, freshHadoopConf, tempCreds)
-        hadoopUtil.obtainTokenForHBase(sparkConf, freshHadoopConf, tempCreds)
+        val nearestNextTime = credentialManager.obtainCredentials(freshHadoopConf, tempCreds)
+        require(nearestNextTime > System.currentTimeMillis(),
```
do we really want this to be a require?  This is going to throw an exception. The exception is caught in driverTokenRenewerRunnable and then logged but then sets a timer to call again in an hour.  We don't want to wait an hour if renewal is < current time we want it to renew right now so if we just set timeOfNextRenewal and return it would do that.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ExecutorDelegationTokenUpdater.scala:
```diff
@@ -107,6 +110,14 @@ private[spark] class ExecutorDelegationTokenUpdater(
+    val index = name.lastIndexOf(SparkHadoopUtil.SPARK_YARN_CREDS_COUNTER_DELIM)
+    val slice = name.substring(0, index)
+    val last2index = slice.lastIndexOf(SparkHadoopUtil.SPARK_YARN_CREDS_COUNTER_DELIM)
+    name.substring(last2index).toLong
```
is this giving you the right time?  It looks like its giving you the credentials file name instead of the next time to update.

name is: credentials-UUID-timestamp-1...   slice = credentials-UUID-timestamp, name.substring of that is credentials-UUID.  don't you want name.substring(last2index, index) ?

jerryshao replied:
Sorry about it, I will fix it.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ServiceCredentialProvider.scala:
```diff
@@ -0,0 +1,55 @@
+ * A credential provider for this service, any service want to access by spark through security
+ * way must implement this interface.
+ */
+@DeveloperApi
```
I dont think this should be a developerAPI, we want people to extends this and not worry about it breaking.  Unless your intention is to make it develop for now until we get a little run time on it?

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ServiceCredentialProvider.scala:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * ::DeveloperApi::
+ * A credential provider for this service, any service want to access by spark through security
```
slight wording change: 

A credentials provider for a service. A user must implement this if they need to access a secure service from spark.

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ServiceCredentialProvider.scala:
```diff
@@ -0,0 +1,55 @@
+
+  /**
+   * To decide whether credential is required for this service. By default it based on whether
+   * security is enabled or not.
```
whether Hadoop security is enabled

tgravescs commented:
what all testing have you done on this? Ran on actual cluster with secure hive, hbase, hdfs?  

Test out tokens and renewal intervals to make sure things propogated?

jerryshao committed:
Further refactor the codes according to the comments

SparkQA commented:
**[Test build #62658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62658/consoleFull)** for PR 14065 at commit [`bd04a2b`](https://github.com/apache/spark/commit/bd04a2b9590e59800e1206103d32779d190056ae).

jerryshao commented:
Hi @tgravescs , thanks a lot for your comments. What I did to test is to run on security cluster with hive, hbase and hdfs installed and kerberized:

1. Verify credentials can be obtained in yarn/client properly.
2. Verify if principal and keytab is specified, credentials can be obtained in yarn/client correctly.
3. Verify if principal and keytab is specified, credentials can be renewed and updated in a specific time. I changed the renewal and update parameter from 75%, 80% to 1% and 1.2% to expect fast credential renewal. To see if:
    * Credentials can be renewed in AM in a specific time.
    * Renewed credentials can be written into hdfs with proper name: <credential>-<timestamp>-<integer suffix>.
    * Credentials can be updated in driver and executors correctly.

SparkQA commented:
**[Test build #62658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62658/consoleFull)** for PR 14065 at commit [`bd04a2b`](https://github.com/apache/spark/commit/bd04a2b9590e59800e1206103d32779d190056ae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62658/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -312,15 +312,15 @@ class SparkHadoopUtil extends Logging {
   /**
-   * Start a thread to periodically update the current user's credentials with new delegation
-   * tokens so that writes to HDFS do not fail.
+   * Start a thread to periodically update the current user's credentials with new credentails so
```
nit: credentials misspelled

tgravescs commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala:
```diff
@@ -171,16 +172,30 @@ private[yarn] class AMDelegationTokenRenewer(
+      // If next renewal time is earlier than current time, we set next renewal time to current
+      // time, this will trigger next renewal immediately. Also set next update time to current
+      // time. There still has a gap between token renewal and update will potentially introduce
+      // issue.
```
It might be useful here to log a message here as we don't really expect this to happen.  If it keeps happening it could be we aren't renewing them soon enough.

tgravescs commented:
couple minor nits, otherwise seems fine, but I want to run a couple tests locally. 

@vanzin were all your comments addressed?

vanzin commented:
I'm having trouble finding the bandwidth to look at the updated patch, but it's on my list... there were some replies to my comments that I want to take a closer look at.

jerryshao committed:
Docs change

SparkQA commented:
**[Test build #62915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62915/consoleFull)** for PR 14065 at commit [`dc3be43`](https://github.com/apache/spark/commit/dc3be432c5467d1d6f0fedbcb0cd172ab0fa8bba).

SparkQA commented:
**[Test build #62915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62915/consoleFull)** for PR 14065 at commit [`dc3be43`](https://github.com/apache/spark/commit/dc3be432c5467d1d6f0fedbcb0cd172ab0fa8bba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62915/
Test PASSed.

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -525,16 +534,22 @@ token for the cluster's HDFS filesystem, and potentially for HBase and Hive.
 the tokens needed to access these clusters must be explicitly requested at
 launch time. This is done by listing them in the `spark.yarn.access.namenodes` property.
 
+Also Spark supports interacting with other security services through plug-in mechanism, user must implement
```
Rephrase:

Spark supports integrating with other security-aware services through the Java Services mechanism (see
`java.util.ServiceLoader`). To do that, implementations of `org.apache.spark.deploy.yarn.security.ServiceCredentialProvider` should be made available to Spark by listing their  ames in the corresponding file in the jar's `META-INF/services` directory. These plug-ins can be disabled by setting...

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala:
```diff
@@ -27,39 +27,42 @@ import org.apache.hadoop.security.UserGroupInformation
- * CoarseGrainedScheduledBackend starts up. This method wakes up a thread that logs into the KDC
- * once 75% of the renewal interval of the original delegation tokens used for the container
- * has elapsed. It then creates new delegation tokens and writes them to HDFS in a
+ * Streaming apps can run without interruption while access security services. The
```
accessing

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,96 @@
+
+    // Filter out credentials in which spark.yarn.security.credentials.{service}.enabled is false.
+    providers.filter { p =>
+      sparkConf.getOption(providerEnabledConfig.format(p.serviceName)).map(_.toBoolean)
```
Minor but you can move the `toBoolean` transformation to after the `orElse` and save a tiny bit of code.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,96 @@
+   * @return nearest time of next renewal, Long.MaxValue if all the credentials aren't renewable,
+   *         otherwise the nearest renewal time of any credentials will be returned.
+   */
+  def obtainCredentials(hadoopConf: Configuration, creds: Credentials): Long = {
```
Do you need `hadoopConf` here again? It's already provided in the class's constructor.

jerryshao replied:
Yes, in `AMCredentialRenewer` it uses a `freshHadoopConf` to obtain credentials instead of the one passed from constructor. So here I added this parameter.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,96 @@
+    }.foldLeft(Long.MaxValue)(math.min)
+  }
+
+  def credentialRenewer: AMCredentialRenewer = {
```
The method is named as if it were a property, yet it returns something different on every call. It should have a signature that reflects that. e.g. `newCredentialRenewer()`. scaladoc would be nice too, especially to clarify that the caller is responsible for the lifecycle of the returned instance.

Also applies to `credentialUpdater`.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -41,16 +43,18 @@ private[spark] class ExecutorDelegationTokenUpdater(
       override def run(): Unit = Utils.logUncaughtExceptions(updateCredentialsIfRequired())
     }
 
+  @volatile private var timeOfNextUpdate = sparkConf.get(CREDENTIALS_UPDATE_TIME)
```
This is only ever used inside `updateCredentialsIfRequired`, do you need to keep it in a field?

jerryshao replied:
This method `updateCredentialsIfRequired ` will periodically call itself, so it is easy to code to put it as a field.

vanzin replied:
Not sure I understand.

`timeOfNextUpdate` is set in L73, and is used in L82. So it doesn't seem like keeping this state across calls of that method is necessary.

I can see a case where, if the directory where credentials are expected to be stored is empty, then the code skips L73. That seems to be used to "bootstrap" the updater; the first invocation of `updateCredentialsIfRequired` will schedule an update based on the renewal time set in the conf.

I find that behavior a little odd and would prefer if the "bootstrapping" were more explicit, i.e., if there was code to explicitly schedule the first `updateCredentialsIfRequired` with the value of `CREDENTIALS_UPDATE_TIME`, and have the code treat an empty directory the same way as if the more recent file does not match the expected (i.e. L76).

I know this code is not being added as part of your change, but could you do that cleanup?

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -107,8 +110,16 @@ private[spark] class ExecutorDelegationTokenUpdater(
   }
 
+  private def getTimeOfNextUpdateFromFileName(credentialsPath: Path): Long = {
+    val name = credentialsPath.getName
```
I wonder if using `name.split(SparkHadoopUtil.SPARK_YARN_CREDS_COUNTER_DELIM)` wouldn't make this method easier to read.

jerryshao replied:
IIUC credential file name is combined by `credentials-uuid-timestamp-integer`, uuid has `-` in it. Maybe it is safer to use the current way.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HDFSCredentialProvider.scala:
```diff
@@ -0,0 +1,118 @@
+  // if None means no token renewer specified, so cannot get token renewal interval.
+  private var tokenRenewalInterval: Option[Long] = null
+
+  override val  serviceName: String = "hdfs"
```
nit: too many spaces after `val`

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HDFSCredentialProvider.scala:
```diff
@@ -0,0 +1,118 @@
+
+private[security] class HDFSCredentialProvider extends ServiceCredentialProvider with Logging {
+
+  private val sparkConf = new SparkConf()
```
Hmm, not sure I like this. Feels like this should be an argument to `obtainCredentials`. Or maybe you should have some kind of explicit initialization method. But code like this makes it hard to mock configuration during tests, for example.

One idea is to modify `isCredentialRequired` so that it's an actual initialization method that returns whether the provider should be active.

jerryshao replied:
I would like to add this `sparkConf` as an argument to `obtainCredentials`. `isCredentialRequired` would be better not to do initialization things, since it will be called every time before obtaining credentials.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HDFSCredentialProvider.scala:
```diff
@@ -0,0 +1,118 @@
+    delegTokenRenewer
+  }
+
+  private def nnsToAccess(sparkConf: SparkConf): Set[Path] = {
```
This seems to only be called from a single place, might as well inline it.

vanzin replied:
(I see it's used in the tests too, let me read that code first.)

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ServiceCredentialProvider.scala:
```diff
@@ -0,0 +1,51 @@
+   * To decide whether credential is required for this service. By default it based on whether
+   * Hadoop security is enabled.
+   */
+  def isCredentialRequired(hadoopConf: Configuration): Boolean = {
```
Aside from my previous comment to maybe re-purpose this method, it should be `credentialsRequired` (or `areCredentialsRequired` but that looks a bit weird in code).

vanzin commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManagerSuite.scala:
```diff
@@ -0,0 +1,153 @@
+  test("plug in customized credential provider") {
+    credentialManager = new ConfigurableCredentialManager(sparkConf, hadoopConf)
+
+    credentialManager.getServiceCredentialProvider("test") should not be (None)
```
minor but this is implicitly tested by the previous test. maybe remove the check for the "test" provider from there.

vanzin commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManagerSuite.scala:
```diff
@@ -0,0 +1,153 @@
+    nextRenewal should be (testCredentialProvider.timeOfNextTokenRenewal)
+  }
+
+  test("Obtain tokens For HiveMetastore") {
```
nit: all test descriptions start with lower case, so should this one (and the next)

vanzin commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/security/HDFSCredentialProviderSuite.scala:
```diff
@@ -0,0 +1,106 @@
+
+  test("check access nns empty") {
+    val sparkConf = new SparkConf()
+      .set("spark.yarn.access.namenodes", "")
```
I don't see a whole lot of value in these tests; you're basically testing the config parsing code, which is already tested thoroughly in other places. There's no custom parsing going on the HDFS provider code, aside from the transformation from String to Path.

If you could instead test that the HDFS provider were actually getting tokens for all NNs that would be great, but I'm not sure you can do that easily with the current test infrastructure (e.g. you'd have to somehow be able to mock the `FileSystem` instance to see if the interesting methods were called).

jerryshao replied:
these bunch of tests existed in the original code, I just moved it to `HDFSCredentialProviderSuite`. I can delete them if you think it is not so necessary.

vanzin commented:
I still see a few, mostly minor, issues. I think the provider initialization should be cleaned up to avoid things like the hacky code in the HDFS provider.

jerryshao committed:
Further address the comments

SparkQA commented:
**[Test build #63068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63068/consoleFull)** for PR 14065 at commit [`ac0062c`](https://github.com/apache/spark/commit/ac0062cf9d8a6f48d03fe335ba2ea7ec5f00e8e7).

SparkQA commented:
**[Test build #63068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63068/consoleFull)** for PR 14065 at commit [`ac0062c`](https://github.com/apache/spark/commit/ac0062cf9d8a6f48d03fe335ba2ea7ec5f00e8e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63068/
Test PASSed.

jerryshao commented:
@vanzin ,  I did refactoring on some interfaces, especially for `obtainCredentials` method, and the implementation of `HDFSCredentialProvider` and `HiveCredential`, would you please help to review again? Thanks a lot.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -338,15 +312,15 @@ class SparkHadoopUtil extends Logging {
-   * Start a thread to periodically update the current user's credentials with new delegation
-   * tokens so that writes to HDFS do not fail.
+   * Start a thread to periodically update the current user's credentials with new credentials so
+   * that access to security service do not fail.
```
nit: s/security/secured; "does not fail"

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -461,15 +461,24 @@ To use a custom metrics.properties for the application master and executors, upd
+  By default, credentials for all supported services are retrieved when those services are
   configured, but it's possible to disable that behavior if it somehow conflicts with the
-  application being run.
+  application being run. For the details please see
```
nit: "For further details, please..."

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -461,15 +461,24 @@ To use a custom metrics.properties for the application master and executors, upd
+  [Running in a Secure Cluster](running-on-yarn.html#running-in-a-secure-cluster)
   <p/>
-  Currently supported services are: <code>hive</code>, <code>hbase</code>
+  Currently supported services are: <code>hdfs</code>, <code>hive</code>, <code>hbase</code>
```
This sentence is now a little misleading; should probably either not mention this, or say that these are the built-in services.

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -461,15 +461,24 @@ To use a custom metrics.properties for the application master and executors, upd
+  </td>
+</tr>
+<tr>
+  <td><code>spark.yarn.security.tokens.${service}.enabled</code></td>
```
I'd just remove deprecated configs from the documentation.

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -525,16 +534,23 @@ token for the cluster's HDFS filesystem, and potentially for HBase and Hive.
+Spark supports integrating with other security-aware services through Java Services mechanism (see
+`java.util.ServiceLoader`). To do that, implementations of `org.apache.spark.deploy.yarn.security.ServiceCredentialProvider`
+ should be available to Spark by listing their names in the corresponding file in the jar's
+ `META-INF/services` directory. These plug-ins can be disabled by setting
```
nit: there's a leading space in this line and the next

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala:
```diff
@@ -27,39 +27,42 @@ import org.apache.hadoop.security.UserGroupInformation
- * CoarseGrainedScheduledBackend starts up. This method wakes up a thread that logs into the KDC
- * once 75% of the renewal interval of the original delegation tokens used for the container
- * has elapsed. It then creates new delegation tokens and writes them to HDFS in a
+ * Streaming apps can run without interruption while accessing security services. The
```
nit: s/security/secured

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala:
```diff
@@ -27,39 +27,42 @@ import org.apache.hadoop.security.UserGroupInformation
+ * scheduleLoginFromKeytab method is called on the AM to get the new credentials.
+ * This method wakes up a thread that logs into the KDC
+ * once 75% of the renewal interval of the original credentials used for the container
+ * has elapsed. It then obtain new credentials and writes them to HDFS in a
```
nit: obtains

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/ConfigurableCredentialManager.scala:
```diff
@@ -0,0 +1,105 @@
+ * Also each credential provider is controlled by
+ * spark.yarn.security.credentials.{service}.enabled, it will not be loaded in if set to false.
+ */
+private[yarn] final class ConfigurableCredentialManager (
```
nit: no space before '('

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/HiveCredentialProvider.scala:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  override def credentialsRequired(hadoopConf: Configuration): Boolean = {
+    if (!UserGroupInformation.isSecurityEnabled) {
```
`UserGroupInformation.isSecurityEnabled && hiveConf...`

vanzin commented:
LGTM, just pointed out one thing that I think would make the code easier to follow.

jerryshao committed:
Style and typo fix

SparkQA commented:
**[Test build #63200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63200/consoleFull)** for PR 14065 at commit [`127d85e`](https://github.com/apache/spark/commit/127d85ed54f057581a35c88fc7f85e1b8e13de38).

SparkQA commented:
**[Test build #63200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63200/consoleFull)** for PR 14065 at commit [`127d85e`](https://github.com/apache/spark/commit/127d85ed54f057581a35c88fc7f85e1b8e13de38).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63200/
Test PASSed.

vanzin commented:
@jerryshao do you plan to address the remaining feedback? (https://github.com/apache/spark/pull/14065#discussion_r73423485)

jerryshao commented:
@vazin, looks like I missed that comment, I will address that today.

jerryshao committed:
Simplify the credential update logic

SparkQA commented:
**[Test build #63349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63349/consoleFull)** for PR 14065 at commit [`32c16c0`](https://github.com/apache/spark/commit/32c16c06bd54eb26abc97c4fe047dd2a9af0c4ad).

SparkQA commented:
**[Test build #63349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63349/consoleFull)** for PR 14065 at commit [`32c16c0`](https://github.com/apache/spark/commit/32c16c06bd54eb26abc97c4fe047dd2a9af0c4ad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63349/
Test PASSed.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
     }
 
-  @volatile private var timeOfNextUpdate = sparkConf.get(CREDENTIALS_UPDATE_TIME)
+  /** Start the credentail updater thread periodically to get the credentials */
```
"credential"

Also the whole comment sounds redundant. Just: "Start the credential updater task."

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
 
-  @volatile private var timeOfNextUpdate = sparkConf.get(CREDENTIALS_UPDATE_TIME)
+  /** Start the credentail updater thread periodically to get the credentials */
+  def bootstrap(): Unit = {
```
I'd call this `startUpdater` or just `start`.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
-  @volatile private var timeOfNextUpdate = sparkConf.get(CREDENTIALS_UPDATE_TIME)
+  /** Start the credentail updater thread periodically to get the credentials */
+  def bootstrap(): Unit = {
+    val bootstrapTime = sparkConf.get(CREDENTIALS_RENEWAL_TIME)
```
Similarly, I'd call this `startTime` or `runTime`.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
+    val bootstrapTime = sparkConf.get(CREDENTIALS_RENEWAL_TIME)
+    val remainingTime = bootstrapTime - System.currentTimeMillis()
+    if (remainingTime <= 0) {
+      // We just checked for new credentials but none were there, wait a minute and retry.
```
This comment is now wrong, since you haven't checked for anything yet.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
         credentialsFilePath.getName, SparkHadoopUtil.SPARK_YARN_CREDS_TEMP_EXTENSION)
-        .lastOption.foreach { credentialsStatus =>
+        .lastOption.map { credentialsStatus =>
         val suffix = SparkHadoopUtil.get.getSuffixForCredentialsPath(credentialsStatus.getPath)
```
Doesn't look like your fault, but this whole block code is indented incorrectly. Should be one level deeper.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
+
+          val remainingTime =
+            getTimeOfNextUpdateFromFileName(credentialsStatus.getPath) - System.currentTimeMillis()
+          if (remainingTime <= 0) 60L * 1000 else remainingTime
```
`TimeUnit.HOUR.toMillis(1)`

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
-            "credentials yet, will check again in an hour.")
-          credentialUpdater.schedule(credentialUpdaterRunnable, 1, TimeUnit.HOURS)
-          return
+          // If current credential file is elder than expected, sleep 1 hour and check again.
```
"older"

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
-          credentialUpdater.schedule(credentialUpdaterRunnable, 1, TimeUnit.HOURS)
-          return
+          // If current credential file is elder than expected, sleep 1 hour and check again.
+          3600L * 1000
```
`TimeUnit.HOUR.toMillis(1)`

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
-          credentialUpdaterRunnable, remainingTime, TimeUnit.MILLISECONDS)
+      }.getOrElse {
+        // Wait for 1 minute to check again if there's no credential file currently
+        60L * 1000
```
`TimeUnit.MINUTE.toMillis(1)`

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala:
```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
       case NonFatal(e) =>
         logWarning("Error while trying to update credentials, will try again in 1 hour", e)
-        credentialUpdater.schedule(credentialUpdaterRunnable, 1, TimeUnit.HOURS)
+        3600L * 1000
```
`TimeUnit...` you get the gist.

jerryshao committed:
Change the codes according the comments

SparkQA commented:
**[Test build #63405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63405/consoleFull)** for PR 14065 at commit [`71df428`](https://github.com/apache/spark/commit/71df4281404907e2143f1d23bf2a4ec606ab6ef2).

SparkQA commented:
**[Test build #63405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63405/consoleFull)** for PR 14065 at commit [`71df428`](https://github.com/apache/spark/commit/71df4281404907e2143f1d23bf2a4ec606ab6ef2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63405/
Test PASSed.

vanzin commented on docs/running-on-yarn.md:
````diff
@@ -525,16 +524,23 @@ token for the cluster's HDFS filesystem, and potentially for HBase and Hive.
+credential provider.
+
 ```
 spark.yarn.access.namenodes hdfs://ireland.example.org:8020/,hdfs://frankfurt.example.org:8020/
````
Hmm, it seems like this example should be before the paragraph you're adding.

vanzin commented:
Looks fine. There are some possible enhancements (e.g. what looks like some code repetition in the HDFS provider, neither Hive nor HBase return a token renewal time, etc) but those can be done separately.

@tgravescs did you have any remaining comments?

tgravescs commented:
all my original comments were addressed and I won't have time to do another review until next week so I'm good with it if you are.

jerryshao committed:
Change the doc

SparkQA commented:
**[Test build #63516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63516/consoleFull)** for PR 14065 at commit [`bce8cd6`](https://github.com/apache/spark/commit/bce8cd6a5796fc95a432c169c8c40bea552382f0).

SparkQA commented:
**[Test build #63516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63516/consoleFull)** for PR 14065 at commit [`bce8cd6`](https://github.com/apache/spark/commit/bce8cd6a5796fc95a432c169c8c40bea552382f0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63516/
Test PASSed.

vanzin commented:
Merging to master.
